components/merkledb/benches/schema_patterns.rs

```diff
@@ -0,0 +1,329 @@
+    cold_index: ProofMapIndex<T::Base, u64, Hash>,
+    cold_group: Group<T, u64, ProofListIndex<T::Base, u64>>,
+    // Accessed once per ~`COLD_DIVISOR` transactions.
+    other_cold_index: KeySetIndex<T::Base, u64>,
+}
+
+impl<T: Access> EagerSchema<T> {
+    fn new(access: T) -> Self {
+        Self {
+            transactions: FromAccess::from_access(access.clone(), "transactions".into()).unwrap(),
+            hot_index: FromAccess::from_access(access.clone(), "hot_index".into()).unwrap(),
+            hot_group: FromAccess::from_access(access.clone(), "hot_group".into()).unwrap(),
+            cold_index: FromAccess::from_access(access.clone(), "cold_index".into()).unwrap(),
+            cold_group: FromAccess::from_access(access.clone(), "cold_group".into()).unwrap(),
+            other_cold_index: FromAccess::from_access(access.clone(), "other_cold_index".into())
+                .unwrap(),
+        }
+    }
+}
+
+impl<T: Access> EagerSchema<T>
+where
+    T::Base: RawAccessMut,
+{
+    fn execute(&mut self, transaction: &Transaction) {
+        self.transactions
+            .put(&transaction.object_hash(), *transaction);
+
+        // Access hot index and group a few times.
+        for &divisor in DIVISORS {
+            let group_id = transaction.value % divisor;
+            let mut list_in_group = self.hot_group.get(&group_id);
+            list_in_group.push(transaction.value);
+            self.hot_index.put(&group_id, list_in_group.object_hash());
+
+            // Cold index / group are accessed only a fraction of the time.
+            if group_id == 0 {
+                let cold_group_id = transaction.value % COLD_DIVISOR;
+                let mut list_in_group = self.cold_group.get(&cold_group_id);
+                list_in_group.push(transaction.value);
+                self.cold_index
+                    .put(&cold_group_id, list_in_group.object_hash());
+            }
+        }
+
+        if transaction.value % COLD_CHANCE == 0 {
+            self.other_cold_index.insert(transaction.value);
+        }
+    }
+}
+
+struct EagerStyle;
+
+impl ExecuteTransaction for EagerStyle {
+    fn execute<T: Access>(fork: T, transaction: &Transaction)
+    where
+        T::Base: RawAccessMut,
+    {
+        let mut schema = EagerSchema::new(fork);
+        schema.execute(transaction);
+    }
+}
+
+struct LazySchema<T: Access> {
+    transactions: MapIndex<T::Base, Hash, Transaction>,
+    hot_index: ProofMapIndex<T::Base, u64, Hash>,
+    hot_group: Group<T, u64, ProofListIndex<T::Base, u64>>,
+    cold_index: Lazy<T, ProofMapIndex<T::Base, u64, Hash>>,
+    // groups are already lazy
+    cold_group: Group<T, u64, ProofListIndex<T::Base, u64>>,
+    other_cold_index: Lazy<T, KeySetIndex<T::Base, u64>>,
+}
+
+impl<T: Access> LazySchema<T> {
+    fn new(access: T) -> Self {
+        Self {
+            transactions: FromAccess::from_access(access.clone(), "transactions".into()).unwrap(),
```

Comment:
One little thought: what if we'd introduced a little helper method to reduce visual code complexity?

Reply:
I had in mind that a derive macro would take care of this. (Now that I think of it, it could make sense to implement the derive macro for `FromAccess` and have a method `FromAccess::<T>::create_here(T)` with a default implementation.) Does it make sense to defer this as a separate clean-up task?